### PR TITLE
feat(core): Min-max versions of histogram quantile and fraction for exp histograms

### PR DESCRIPTION
**Pull request by vishramachandran**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

Use min and max when available to calculate better quantiles and fractions.

Updated perf test results for quantiles with exponential histograms are shown below. There is a proportional degradation of performance as the number of buckets increase. Thus, exponential histograms will benefit from pre-aggregation when a large number of time series is involved.   **Note these results supersede the results in earlier exponential histogram PR.**

```
32 buckets without profiling
[info] Benchmark                                                          Mode  Cnt    Score   Error  Units
[info] Base2ExponentialHistogramQueryBenchmark.histQuantileRangeQueries  thrpt    2  404.993          ops/s

64 buckets without profiling
[info] Benchmark                                                          Mode  Cnt    Score   Error  Units
[info] Base2ExponentialHistogramQueryBenchmark.histQuantileRangeQueries  thrpt    2  252.478          ops/s

128 buckets without profiling
[info] Benchmark                                                          Mode  Cnt    Score   Error  Units
[info] Base2ExponentialHistogramQueryBenchmark.histQuantileRangeQueries  thrpt    2  234.001          ops/s

160 buckets without profiling
[info] Benchmark                                                          Mode  Cnt    Score   Error  Units
[info] Base2ExponentialHistogramQueryBenchmark.histQuantileRangeQueries  thrpt    2  219.758          ops/s

200 buckets without profiling
[info] Benchmark                                                          Mode  Cnt    Score   Error  Units
[info] Base2ExponentialHistogramQueryBenchmark.histQuantileRangeQueries  thrpt    2  137.804          ops/s

256 buckets without profiling
[info] Benchmark                                                          Mode  Cnt    Score   Error  Units
[info] Base2ExponentialHistogramQueryBenchmark.histQuantileRangeQueries  thrpt    2  119.017          ops/s
```

Fitted Linear Regression:
image

JMH Benchamark CPU Profile 
image

### CODE REVIEW

File: query/src/main/scala/filodb/query/exec/MultiSchemaPartitionsExec.scala [hunk 1]
```diff
@@ -128,10 +128,10 @@ final case class MultiSchemaPartitionsExec(queryContext: QueryContext,
             // This code is responsible for putting exact IDs needed by any range functions.
             val colIDs1 = getColumnIDs(sch, newColName.toSeq, rangeVectorTransformers)
 
-            val colIDs = isMaxMinColumnsEnabled(maxMinTenantFilter) match {
-              case true => addIDsForHistMaxMin(sch, colIDs1)
-              case _ => colIDs1
-            }
+            val colIDs = if (sch.data.columns.exists(_.name == "min") &&
```
Comment (sandeep6189) [inline query/src/main/scala/filodb/query/exec/MultiSchemaPartitionsExec.scala:131]: Thank you for adding this check @vishramachandran

File: query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala [hunk 2]
```diff
@@ -366,6 +366,52 @@ class SumAndMaxOverTimeFuncHD(maxColID: Int) extends ChunkedRangeFunction[Transi
...
+    // At least one sample is present
+    if (startRowNum <= endRowNum) {
+      hFunc.addTimeChunks(valueVectorAcc, valueVector, valueReader, startRowNum, endRowNum)
+
+      // Get valueVector/reader for max column
+      val maxVectAcc = info.vectorAccessor(maxColId)
+      val maxVectPtr = info.vectorAddress(maxColId)
+      maxFunc.addTimeChunks(maxVectAcc, maxVectPtr, bv.DoubleVector(maxVectAcc, maxVectPtr), startRowNum, endRowNum)
+
+      // Get valueVector/reader for max column
```
Comment (sandeep6189) [inline query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala:406]: nitpick: update comment `min` column

**Review by sandeep6189 — APPROVED** [review]

**Comment by sandeep6189** [comment]
thank you for adding unit test @vishramachandran

File: memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala [hunk 3]
```diff
@@ -62,28 +62,36 @@ trait Histogram extends Ordered[Histogram] {
...
    * Calculates histogram quantile based on bucket values using Prometheus scheme (increasing/LE)
    */
-  def quantile(q: Double): Double = {
+  def quantile(q: Double,
+               min: Double = 0, // negative observations not supported yet
+               max: Double = Double.PositiveInfinity): Double = {
     val result = if (q < 0) Double.NegativeInfinity
     else if (q > 1) Double.PositiveInfinity
-    else if (numBuckets < 2) Double.NaN
+    else if (numBuckets < 2 || topBucketValue <= 0) Double.NaN
```
[the remark this answers is not in the document]
Comment (vishramachandran) [reply memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala:70]: We don't support negative observations, hence NaN

File: memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala [hunk 4]
```diff
@@ -62,28 +62,36 @@ trait Histogram extends Ordered[Histogram] {
...
-      else {
-        // interpolate quantile within le bucket
-        var (bucketStart, bucketEnd, count) = (0d, bucketTop(b), bucketValue(b))
-        if (b > 0) {
-          bucketStart = bucketTop(b-1)
-          count -= bucketValue(b-1)
-          rank -= bucketValue(b-1)
-        }
+      if (bucket == numBuckets-1 && bucketTop(numBuckets - 1).isPosInfinity) {
+        return bucketTop(numBuckets-2)
```
[the remark this answers is not in the document]
Comment (vishramachandran) [reply memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala:87]: This is to be backward compatible with case where max is not available (and defaults to PosInfinity). 
We can add additional edge case to include the situation where max is available. Let me look into it.

File: memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala [hunk 5]
```diff
@@ -62,28 +62,36 @@ trait Histogram extends Ordered[Histogram] {
...
+      } else if (bucket == 0 && bucketTop(0) <= 0) {
+        return bucketTop(0) // zero or negative bucket
```
[the remark this answers is not in the document]
Comment (vishramachandran) [reply memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala:89]: Is that a valid case ? Can bucket == 0  (which means there are observations in this bucket) and min > bucketTop(0) 

In any case, we don't support negative observations yet, so this should be good.

File: memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (vishramachandran) [reply memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala:87]: Fixing to `if (bucket == numBuckets-1 && bucketEnd.isPosInfinity)`

**Review by sandeep6189 — APPROVED** [review]

**Pull request merged**